Nuttall window

alvgaona opened the pull request:
Relates #85366

alvgaona committed:
Add Nuttall window

alvgaona committed:
Remove linebreak

alvgaona commented:
@lezcano 😄

lezcano commented on torch/signal/windows/windows.py:
```diff
@@ -759,3 +761,64 @@ def general_hamming(M,
+
+@_add_docstr(
+    r"""
+Computes the minimum 4-term Blackman-Harris window according to Nuttall.
```
Formula?

alvgaona replied:
Sure, it's a sum of cosines. I'll add it.

lezcano commented on torch/signal/windows/windows.py:
```diff
@@ -759,3 +761,64 @@ def general_hamming(M,
+    {M}
+
+Keyword args:
+    alpha (float, optional): the window coefficient. Default: 0.54.
```
This function does not take an alpha parameter?

lezcano commented on torch/signal/windows/windows.py:
```diff
@@ -759,3 +761,64 @@ def general_hamming(M,
+    {device}
+    {requires_grad}
+
+Returns:
+    {return}
```
I was thinking when looking at the docs, do you think we should just remove the `Returns` field from all the windows? It adds quite a bit of vertical space (two new fields) and it provides little-to-none new info.

lezcano commented on torch/testing/_internal/opinfo/definitions/signal.py:
```diff
@@ -824,4 +824,57 @@ def make_signal_windows_opinfo(
+        sample_inputs_func=sample_inputs_window,
+        reference_inputs_func=reference_inputs_window,
+        error_inputs_func=error_inputs_window,
+        skips=(
+            DecorateInfo(
+                unittest.expectedFailure,
+                "TestDecomp",
+                "test_comprehensive",
+                dtypes=[torch.float16],
+                device_type="cpu",
+            ),
```
I think we should disallow now that we can `float16` and `bfloat16` dtypes as they have all these weird edge cases. We can always enable them later again if people ask for them. What do you think?

lezcano reviewed (approved):
I left a few small comments here and there, but otherwise it looks good. Thank you for the continued effort pushing this module forward!

alvgaona committed:
Address comments (docs & types)

- Add Nuttall window formula
- Remove float16 and bfloat16 types from the window API

alvgaona commented:
> I left a few small comments here and there, but otherwise it looks good. Thank you for the continued effort pushing this module forward!

Thanks. 😄

alvgaona committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into nutall-window

alvgaona committed:
squash

lezcano commented on torch/signal/windows/windows.py:
```diff
@@ -759,3 +732,71 @@ def general_hamming(M,
+Computes the minimum 4-term Blackman-Harris window according to Nuttall.
+
+.. math::
+    w_n = 1 - 0.3635819 \cos{(z)} + 0.4891775 \cos{(2z)} - 0.1365995 \cos{(3z)} + 0.0106411 \cos{(4z)}
```
```suggestion
    w_n = \sum_{j=0}^{M - 1} 1 - 0.3635819 \cos{(z_j)} + 0.4891775 \cos{(2z_j)} - 0.1365995 \cos{(3z_j)} + 0.0106411 \cos{(4z_j)}
```

lezcano replied:
nb. If it doesn't fit the formula in the width of the page, perhaps you can trim the last 1 or 2 digits from the decimal expansion of the coefficients?

alvgaona replied:
Trimmed 2 and now it fits. Let's see what the build renders.

lezcano commented on torch/signal/windows/windows.py:
```diff
@@ -759,3 +732,71 @@ def general_hamming(M,
+.. math::
+    w_n = 1 - 0.3635819 \cos{(z)} + 0.4891775 \cos{(2z)} - 0.1365995 \cos{(3z)} + 0.0106411 \cos{(4z)}
+
+where
+
+.. math::
+    z = \frac{2 \pi j}{M}, j = 0 \cdots M - 1
```
```suggestion
where ``z_j = 2 π j / M``.
```
to save some vertical space.

lezcano reviewed (commented):
Thank you for the clean-up! I think the formula has a couple errata though. Please have a look and tell me what you think.

alvgaona committed:
Address docs comments

alvgaona committed:
Remove sum

alvgaona commented on torch/signal/windows/windows.py:
```diff
@@ -759,3 +732,67 @@ def general_hamming(M,
+Computes the minimum 4-term Blackman-Harris window according to Nuttall.
+
+.. math::
+    w_n = 1 - 0.36358 \cos{(z_n)} + 0.48917 \cos{(2z_n)} - 0.13659 \cos{(3z_n)} + 0.01064 \cos{(4z_n)}
```
@lezcano this formula computes each window timestep, no need to add `sum`.

lezcano replied:
You are completely right. My bad!

lezcano reviewed (approved):
In it goes!

lezcano commented:
@pytorchbot merge -f "failure unrelated"
